compiler/src/main/java/com/squareup/anvil/compiler/codegen/Contribution.kt

```diff
@@ -0,0 +1,176 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.anvil.annotations.ContributesTo
+import com.squareup.anvil.annotations.internal.InternalBindingMarker
+import com.squareup.anvil.compiler.BINDING_MODULE_SUFFIX
+import com.squareup.anvil.compiler.MULTIBINDING_MODULE_SUFFIX
+import com.squareup.anvil.compiler.internal.capitalize
+import com.squareup.anvil.compiler.internal.createAnvilSpec
+import com.squareup.anvil.compiler.internal.reference.generateClassName
+import com.squareup.anvil.compiler.internal.reference.generateClassNameString
+import com.squareup.anvil.compiler.internal.safePackageString
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asClassName
+import com.squareup.kotlinpoet.joinToCode
+import dagger.Binds
+import dagger.Module
+import dagger.Provides
+import dagger.multibindings.IntoMap
+import dagger.multibindings.IntoSet
+import java.util.Comparator
+
+internal sealed interface Contribution {
+  val origin: ClassName
+  val scope: ClassName
+  val isObject: Boolean
+  val boundType: ClassName
+  val replaces: List<ClassName>
+  val qualifier: QualifierData?
+  val bindingModuleNameSuffix: String
+
+  // Can't be private until we move off jvmTarget 8
+  fun generateBindingModule(): TypeSpec {
+    val contribution = this
+    // Combination name of origin, scope, and boundType
+    val suffix = "As" +
+      contribution.boundType.generateClassNameString().capitalize() +
+      "To" +
+      contribution.scope.generateClassNameString().capitalize() +
+      bindingModuleNameSuffix
```

Comment:
Could we break this up with some underscores?

Reply:
Yeah we could I was less worried about readability given this is never accessed directly by users and trying to minimize chars to avoid too-long class names